File: src/15장/15-3-3.py
from collections import deque

N,M,V=map(int,input().split())
vertex=[[0]*(N+1) for i in range(N+1)]
for i in range(M):
    a,b = map(int,input().split())
    vertex[a][b]=vertex[b][a]=1
visit=[0]*(N+1)

def dfs(V):
    visit[V]=1 
    print(V, end=' ')
    for i in range(1,N+1):
        if(visit[i]==0 and vertex[V][i]==1):
            dfs(i)

def bfs(V):
    visit[V]=0 
    queue=deque() 
    queue.append(V)
    while queue:
        V=queue.popleft()
        print(V, end=' ')
        for i in range(1, N+1):
            if(visit[i]==1 and vertex[V][i]==1):
                queue.append(i)
                visit[i]=0

dfs(V)
print()
bfs(V)

File: src/15장/15-3-3.java
import java.util.*;
import java.io.*;

public class Main {
	public static int[][] vertex;
	public static boolean[] visit;
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		int m = sc.nextInt();
		int v = sc.nextInt();
		vertex = new int[n+1][n+1];
		for(int i=1;i<=m;i++) {
			int a = sc.nextInt();
			int b = sc.nextInt();
			vertex[a][b] = 1;
			vertex[b][a] = 1;
		}
		visit = new boolean[n+1];
		dfs(v); 
		System.out.println();
		visit = new boolean[n+1];
		bfs(v); 

		
	}
	public static void dfs(int v) {
		visit[v] = true;
		System.out.print(v+ " ");
		if(v == vertex.length) {
			return;
		}
		for(int a=1;a<vertex.length;a++) {
			if(vertex[v][a] == 1 && visit[a] == false) {
				dfs(a);
			}
		}
			
	}
	public static void bfs(int v) {
		Queue<Integer> que = new LinkedList<Integer>(); 
		que.add(v);
		visit[v] = true;
 		System.out.print(v+ " ");
		while(!que.isEmpty()) {
			int temp = que.peek();
			que.poll();
			for(int i=1; i<vertex.length;i++) {
				if(vertex[temp][i] ==1 && visit[i] == false) {
					que.add(i);
					visit[i] = true;
					System.out.print(i+ " ");
				}
			}
		}
	}
}